ccdaservice/ccda_gateway.php

```diff
@@ -7,140 +7,155 @@
  * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
  */
 
 //authenticate for portal or main- never know where it gets used
 
 // Will start the (patient) portal OpenEMR session/cookie.
-require_once(dirname(__FILE__) . "/../src/Common/Session/SessionUtil.php");
+
+use OpenEMR\Common\Csrf\CsrfUtils;
+
+require_once(__DIR__ . "/../src/Common/Session/SessionUtil.php");
 OpenEMR\Common\Session\SessionUtil::portalSessionStart();
 
+$sessionAllowWrite = true;
 if (isset($_SESSION['pid']) && isset($_SESSION['patient_portal_onsite_two'])) {
     $pid = $_SESSION['pid'];
     $ignoreAuth = true;
-    require_once(dirname(__FILE__) . "/../interface/globals.php");
+    require_once(__DIR__ . "/../interface/globals.php");
     define('IS_DASHBOARD', false);
     define('IS_PORTAL', $_SESSION['pid']);
 } else {
     OpenEMR\Common\Session\SessionUtil::portalSessionCookieDestroy();
     $ignoreAuth = false;
-    require_once(dirname(__FILE__) . "/../interface/globals.php");
+    require_once(__DIR__ . "/../interface/globals.php");
     if (!isset($_SESSION['authUserID'])) {
         $landingpage = "index.php";
         header('Location: ' . $landingpage);
         exit;
     }
-
     define('IS_DASHBOARD', $_SESSION['authUserID']);
     define('IS_PORTAL', false);
 }
 
-// give me something to do.
-$dowhat = isset($_REQUEST['action']) ? $_REQUEST['action'] : '';
-if ($dowhat && $GLOBALS['ccda_alt_service_enable'] > 0) {
-    if (!checkService()) { // woops, try again
-        if (!checkService()) { // thats 10 seconds of wasted time.
-            die("Document service start failed. Click back to return home."); // nuts! give up
-        }
-    }
-} else {
-    // maybe next time
-    die("Cda generation service turned off: Verify in Administration->Globals! Click back to return home."); // Die an honorable death!!
+if (!CsrfUtils::verifyCsrfToken($_GET["csrf_token_form"])) {
+    CsrfUtils::csrfNotVerified();
 }
 
-//eventually below will qualify what document to fetch
-$parameterArray = array();
-$parameterArray ['encounter'];
-$parameterArray ['combination'] = $pid;
-$parameterArray ['components']; // = 'progress_note|consultation_note|continuity_care_document|diagnostic_image_reporting|discharge_summary|history_physical_note|operative_note|procedure_note|unstructured_document';
-$parameterArray ['sections']; // = 'allergies|medications|problems|immunizations|procedures|results|plan_of_care|vitals|social_history|encounters|functional_status|referral|instructions';
-$parameterArray ['downloadccda'] = 1;
-$parameterArray ['sent_by'];
-$parameterArray ['send'];
-$parameterArray ['view'] = 1;
-$parameterArray ['recipients'] = 'patient'; // emr_direct or hie else if not set $_SESSION['authUserID']
-$parameterArray [0][6] = $_SESSION ['portal_username']; // set to an onsite portal user
+$dowhat = $_REQUEST['action'] ?? '';
+if ((!$GLOBALS['ccda_alt_service_enable']) > 0) {
```

Comment:
this seems like an odd comparison where going to see if a boolean (true or false) is greater than 0

Reply:
is this meant to be:
```
if (!empty($GLOBALS['ccda_alt_service_enable']) && ($GLOBALS['ccda_alt_service_enable'] > 0)) {
```
if so, just let me know, and I can add this change to a branch/PR I am working on

Reply:
shhh! not sure what I was thinking. prob about fishing!
should be
`$GLOBALS['ccda_alt_service_enable'] > 1`
1 is for CCM only

Reply:
done :)

Reply:
thank you...